av/audio/frame.pyx

```diff
@@ -104,21 +106,37 @@ cdef class AudioFrame(Frame):
         )
 
     @staticmethod
-    def from_ndarray(array, format='s16', layout='stereo'):
+    def from_ndarray(array, format=None, layout=None):
         """
         Construct a frame from a numpy array.
+
+        .. note:: The type and number of channels are automatically deduced.
         """
         import numpy as np
 
-        # map avcodec type to numpy type
-        try:
-            dtype = np.dtype(format_dtypes[format])
-        except KeyError:
-            raise ValueError('Conversion from numpy array with format `%s` is not yet supported' % format)
+        if format is not None:
+            logging.warning('deprecated warning : the `format` argument is ignored')
+
+        # map numpy type to avcodec type
+        format = {np_t: av_t for av_t, np_t in format_dtypes.items()}.get(array.dtype.str[-2:], None)
```

Comment:
How can you "autodetect" the format when multiple formats use the same `dtype`?

Reply:
I thought there was no difference between types ending with p and without p.
When you're in mono, it doesn't make a difference. However, in stereo, it does indeed make a difference. It seems to me that the non-planar format is preferable, so that's what is used.